File: chapter6/chapter6.cpp
#include <iostream>
#include <math.h>
#include <set>
#include <vector>
#include <array>
#include <algorithm>
#include <string.h>
#include <cstdlib>
#include <numeric> // accumulate
#include <stdio.h> 
#include <stdlib.h>
#include <sstream>
#include <tuple>
#include <deque>
#include <exception>
#include <memory>
#include <chrono>
#include <mutex>
#include <future>
#include <iomanip>
#include <errno.h>
#include <string.h>
#include "chapter6.h"
extern int errno;


namespace chapter6
{
	template <typename T>
	void SLQueue<T>::enqueue(const T& value)
	{
		SLNode<T> *node = this->slnode(value);
		if (!head) {
			head = node;
		} else {
			tail->next = node;
		}
		tail = node;
	}

	template <typename T>
	void SLQueue<T>::print(SLNode<T> *node)
	{
		for (SLNode<T> *ptr = (!!node) ? node: head ; ptr; ptr = ptr->next) {
			std::cout << ptr->val << '\n';
		}
	}

	template <typename T>
	std::tuple <T, bool> SLQueue<T>::dequeue()
	{
		SLNode<T>* target; 
		T node_value;
		if (head) {
			target = head;
			head = head->next;
			node_value = target->val;
			if (head == NULL) {
				head = NULL;
				tail = NULL;
			}
			delete target;
			return { node_value, true};
		}
		return {NULL, false};
	}

	template <typename T>
	bool SLQueue<T>::contains(const T& value)
	{
		for (SLNode<T>* ptr = head; ptr; ptr = ptr = ptr->next) {
			if (ptr->val == value) {
				return true;
			}
		}
		return false;
	}

	template <typename T>
	bool SLQueue<T>::is_empty()
	{
		return head == NULL;
	}

	template <typename T>
	unsigned SLQueue<T>::size()
	{
		unsigned count = 0;
		for (SLNode<T>* ptr = head; ptr; ptr = ptr->next) {
			++count;
		}
		return count;
	}


	template <typename T>
	SLNode<T> *SLQueue<T>::front () 
	{
		return head;
	}

	template <typename T>
	SLNode<T>* SLQueue<T>::back()
	{
		return tail;
	}

	template <typename T>
	bool SLQueue<T>::compare_queues(SLNode<T>* a, SLNode<T>* b)
	{
		SLNode<T>* ptr_a = a, *ptr_b = b;
		
		while (!!ptr_a || !!ptr_b) {
			
			if (!ptr_a ^ !ptr_b) 
				return false;
			else if (ptr_a->val != ptr_b->val)
				return false;
			
			if (!!ptr_a)
				ptr_a = ptr_a->next;
			
			if (!!ptr_b)
				ptr_b = ptr_b->next;
		
		}
		return true;
	}

	template <typename T>
	void SLQueue<T>::remove_minimums()
	{
		SLNode<T> *ptr, *min_node = head, *prev = head, *target = NULL;
		T min_value;

		// find min
		for (ptr = head; ptr; ptr = ptr->next) {
			if (ptr->val < min_node->val) {
				min_node = ptr;
			}
		}

		ptr = head; 
		// delete min
		if (!!min_node) {
			min_value = min_node->val;
			while (ptr) {
				if (min_value == ptr->val) {
					target = ptr;
					if (target == head) {
						head = head->next;
						ptr = head;
					}	else {
						prev->next = target->next;
						ptr = prev;
					}
					if (target == tail) 
						tail = prev;
					delete target;
				} else {
					prev = ptr;
					ptr = ptr->next;
				}
			}
		}

	}

	template <typename T>
	void SLQueue<T>::remove_last_minimum()
	{
		SLNode<T> *ptr, *min_node = head, *min_node_prev = NULL, *prev = head;

		// find min
		for (ptr = head; ptr; ptr = ptr->next) {
			if (ptr->val <= min_node->val) {
				min_node = ptr;
				min_node_prev = prev;
			}
			prev = ptr;
		}

		if (tail == min_node) {
			if (tail == head) 
				tail = tail->next;
			else 
				tail = min_node_prev;
		}
		
		if (head == min_node) 
			head = head->next;
		else 
			min_node_prev->next = min_node->next;
		

		delete min_node;
	}

	template <typename T>
	SLNode<T> *SLQueue<T>::interleave_queue()
	{
		SLNode<T> *a_prev = NULL, *a = head;
		SLNode<T> *b = head, *b_prev = NULL, *b_node = NULL;
		bool step_pointer = false;

		for (SLNode<T> *ptr = head; ptr; ptr = ptr->next) {
			step_pointer = !step_pointer;
			if (step_pointer) {
				b_prev = b;
				b = b->next;
			}
		}
		b_prev->next = NULL;

		while (!!b) {
			b_node = b;
			b = b->next;
			if (!!a) {
				b_node->next = a->next;
				a->next = b_node;
				a_prev = b_node;
			} else {
				a_prev->next = b_node;
				b = NULL;
			}
		}
		return a;
	}
	
	template <typename T>
	bool SLQueue<T>::is_palindrome(SLQueue<T> &curr_queue)
	{
		SLStack<T> stack;
		std::tuple <T, bool> x_stack, x_queue; 
		unsigned counter = 0;

		// find size 
		while (!curr_queue.is_empty()) {
			counter++;
			stack.push(std::get<0>(curr_queue.dequeue()));
		}

		// throw back into queue
		while (!stack.is_empty()) {
			curr_queue.enqueue(std::get<0>(stack.pop()));
		}

		// copy values in queue to stack
		for (int i = counter; i-- ;) {
			stack.push(curr_queue.front()->val);
			curr_queue.enqueue(std::get<0>(curr_queue.dequeue()));
		}

		//	compare values
		while ( !stack.is_empty() ) {
			x_stack = stack.pop();
			x_queue = curr_queue.dequeue();

			if (!std::get<1>(x_queue) || (std::get<0>(x_stack) != std::get<0>(x_queue))) {
				return false;
			}

			if (std::get<0>(x_stack) == std::get<0>(x_queue)) {
				curr_queue.enqueue(std::get<0>(x_stack));
			}

		}

		return true;
	
	}

	template <typename T>
	void SLQueue<T>::reorder_absolute_queue(SLQueue<T> &queue)
	{
		SLStack<T> stack;			// max values at bottom
		T queue_data;

		while (!queue.is_empty()) {

			queue_data = std::get<0>(queue.dequeue());

			while (!stack.is_empty()) {
				if (queue_data > stack.top()->val) {
					queue.enqueue(std::get<0>(stack.pop()));
				} else {
					break;
				}
			}
			stack.push(queue_data);
		}
	
		while (!stack.is_empty()) {
			queue.enqueue(std::get<0>(stack.pop()));
		}
	}

	template class SLQueue<unsigned>;
	template class SLQueue<signed>;

	template <typename T>
	void SLStack<T>::push(const T& value)
	{
		SLNode<T> *node = this->slnode(value);
		if (this->head == NULL) {
			this->head = node;
			this->tail = node;
		} else {
			node->next = this->head;
			this->head = node;
		}
	}
	


	template<typename T>
	SLNode<T>* SLStack<T>::top()
	{
		return this->head; 
	}

	template<typename T>
	std::tuple<T, bool>  SLStack<T>::pop()
	{
		SLNode<T>* target = NULL;
		T value;
		bool state;

		if (this->head != NULL) {
			target = this->head;
			value = target->val;
			this->head = this->head->next;
			state = true;
		} else {
			value = NULL;
			state = false;
		}

		if (this->head == NULL) {
			this->head = NULL;
			this->tail = NULL;
		}

		if (!!target) {
			delete target;
		}

		return  { value, state };
	}
	
	template <typename T>
	bool SLStack<T>::compare_stacks(SLNode<T>* a, SLNode<T>* b)
	{
		return  this->compare_queues(a, b);
	}

	template <typename T>
	 SLStack<T> SLStack<T>::copy(const  SLStack<T>&  curr_stack)
	{
		SLStack<T> stack;
		SLQueue<T> queue;
		
		for (auto ptr = curr_stack.head; ptr; ptr = ptr->next) 
			stack.push(ptr->val);
	
		while (!stack.is_empty()) 
			queue.enqueue(std::get<0>(stack.dequeue()));

		while (!queue.is_empty()) 
			stack.push(std::get<0>(queue.dequeue()));
	
		return  stack;
	}

	 template <typename T>
	 void SLStack<T>::remove_stack_min(unsigned mode)
	 {
		 SLQueue<T> queue;
		 T min = NULL, stack_data, queue_data;
		 unsigned count = 0;
		 bool state = false;
		 
		 // write to queue (store min)
		 while (!this->is_empty()) {
			 stack_data = std::get<0>(this->pop());
			 if (count == 0 || stack_data < min) {
				 min = stack_data;
				 ++count;
			 }
			 queue.enqueue(stack_data);
		 }

		 while (!queue.is_empty()) {
			 queue_data = std::get<0>(queue.dequeue());
			 /* remove all mins */
			 if (mode == 0) {
				 if (queue_data != min) {
					 this->push(queue_data);
				 }
			 }

			 /* remove first occurence of min*/
			 if (mode == 1) {
				 if (queue_data != min || (state && (queue_data == min))) {
					this->push(queue_data);
				 }
				 state |= (queue_data == min);
			 }

			 /* remove first occurence of min*/
			 if (mode == 2) {
				 if (queue_data != min || count > 0) {
					 this->push(queue_data);
				 }
				 count -= +(queue_data == min);
			 }

		 }

		 // flip stack 
		 while (!this->is_empty()) {
			 stack_data = std::get<0>(this->pop());
			 queue.enqueue(stack_data);
		 }
		 while (!queue.is_empty()) {
			 queue_data = std::get<0>(queue.dequeue());
				 this->push(queue_data);
		 }

	 }
	 
	 template <typename T>
	 void SLStack<T>::remove_newest_min()
	 {
		 remove_stack_min(1);
	 }

	 template <typename T>
	 void  SLStack<T>::remove_oldest_min()
	 {
		 remove_stack_min(2);
	 }

	 template <typename T>
	 void SLStack<T>::partition()
	 {
		 SLQueue<T> queue;
		 T stack_data;
		 bool has_neg = false, has_pos = false;

		 while (!this->is_empty()) {

			 stack_data = std::get<0>(this->pop());
			 
			 if (stack_data <= 0  && has_pos) {
				 while (queue.front()->val < 0 ) {
					 queue.enqueue(std::get<0>(queue.dequeue()));
				 }
			 } 
			 
			 if (stack_data > 0  && has_neg){
				 while (queue.front()->val > 0) {
					 queue.enqueue(std::get<0>(queue.dequeue()));
				 }
			 }

			 queue.enqueue(stack_data);
			 has_neg |= (stack_data <= 0);
			 has_pos |= (stack_data > 0);
		 }

		 while (!queue.is_empty()) {
			 this->push(std::get<0>(queue.dequeue()));
		 }
	 }

	 template <typename T>
	 void SLStack<T>::switch_pairs()
	 {
		 SLQueue<T> queue;
		 std::tuple<T, bool> a, b, c = {NULL, NULL};
		 unsigned count = 0;

		 while (!this->is_empty() ) {
			 count++;
			 queue.enqueue(std::get<0>(this->pop()));
		 }

		 if (count % 2 == 1) {
			 c = queue.dequeue();
			 count--;
		 }
		 
		 for (count -= 2; !queue.is_empty() ; count -= 2) {
			 for (int i = count; i--; ) 
				 queue.enqueue(std::get<0>(queue.dequeue()));
			 a = queue.dequeue();
			 b = queue.dequeue();
			 if (std::get<1>(a)) 
				 this->push(std::get<0>(a));
			 if (std::get<1>(b))
				 this->push(std::get<0>(b));
		 }
		 
		 if (std::get<1>(c)) 
			 this->push(std::get<0>(c));
	 }

	 template <typename T>
	 bool SLStack<T>::is_sorted(SLStack<T> &stack)
	 {
		 SLQueue<T> queue;
		 T data;
		 while (!stack.is_empty()) {
			 data = std::get<0>(stack.pop());
			 if (!queue.is_empty()) {
				 if (data < queue.front()->val)
					 return false;
			 }
			 queue.enqueue(data);
		 }
		 return true;
	 }

	 template <typename T> 
	 void SLStack<T>::mirror(SLStack<T>& stack)
	 {
		 int count = 0;
		 SLQueue<T> queue;
		 T data; 

		 while ( !stack.is_empty() ) {
			 data = std::get<0>(stack.pop());
			 queue.enqueue(data);

			 for (int i = count; i--;) {
				 queue.enqueue(std::get<0>(queue.dequeue()));
			 }
			 queue.enqueue(data);
			 count += 2;
		 }

		 for (int i = count / 2; i--;) {
			queue.enqueue(std::get<0>(queue.dequeue()));
		 }

		 while (!queue.is_empty()) {
			 this->push(std::get<0>(queue.dequeue()));
		 }

	 }


	template class SLStack<unsigned>;
	template class SLStack<signed>;

	template <typename T>
	std::tuple <T, bool> SLQueue_via_Stacks<T>::dequeue()
	{
		return b.pop();
	}

	template <typename T>
	void SLQueue_via_Stacks<T>::enqueue(const T& value)
	{
		move_src_to_dest(b, a); /* b -> a */
		a.push(value);
		move_src_to_dest(a, b); /* a -> b */
	}

	template <typename T>
	void SLQueue_via_Stacks<T>::move_src_to_dest(SLStack<T>& src, SLStack<T>& dest)
	{
		bool is_valid = false;
		T data;
		while (!src.is_empty()) {
			auto tuple = src.pop();
			if (std::get<1>(tuple)) {
				dest.push(std::get<0>(tuple));
			}
		}
	}

	template <typename T>
	void  SLQueue_via_Stacks<T>::print()
	{
		for (SLNode<T>* ptr = b.top(); ptr; ptr = ptr->next) {
			std::cout << "\t data \t " << ptr->val << '\n';
		}
	}

	template class SLQueue_via_Stacks<unsigned>;
	template class SLQueue_via_Stacks<signed>;

	template <typename T>
	SLNode<T>* Dequeue<T>::front()
	{
		return this->front();
	}

	template <typename T>
	SLNode<T>* Dequeue<T>::back()
	{
		return this->back();
	}

	template <typename T>
	void Dequeue<T>::push_back(const T value)
	{
		this->enqueue(value);
	}

	template <typename T>
	void Dequeue<T>::push_front(const T value)
	{
		this->push(value);
	}

	template <typename T>
	std::tuple<T, bool> Dequeue<T>::pop_front()
	{
		return this->pop();
	}

	template <typename T>
	std::tuple<T, bool> Dequeue<T>::pop_back()
	{
		SLNode<T>* header = this->head, * prev = NULL, *ptr;
		T value;

		for (ptr = this->head; (ptr != this->tail) && (ptr != NULL); ptr = ptr->next) {
			prev = ptr;
		}

		if (!this->head) {
			return { NULL, false };
		}

		if (this->head == this->tail) {
			return pop_front();
		}

		if (!!prev) {
			prev->next = prev->next->next;
			value = ptr->val;
			this->tail = prev;
			delete ptr;
			return { value, true };
		}

	}

	template <typename T>
	void Dequeue<T>::print()
	{
		dynamic_cast<SLQueue<T>*>(this)->print();
	}

	template class Dequeue<unsigned>;
	template class Dequeue<signed>;

	template <typename T>
	void CirQueue<T>::shrink(const unsigned &size)
	{
		SLNode<T>* node = head, * prev = head, *target;
		
		for (int i = size; i--;) {
			prev = node; 
			node = node->next;
		}

		if (!!prev) {
			tail = prev;
			target = node;
			prev->next = NULL;

			while (!!node) {
				target = node ;
				node = node->next;
				delete target;
			}

		}
	}

	template <typename T>
	void CirQueue<T>::grow(const unsigned& size)
	{
		SLNode<T>* node = head, *prev = head; 
		
		if (size < 1)
			return;

		if (size < cap) {
			return shrink(size);
		}

		if (size >= cap)
			cap = size;
	}

	template <typename T>
	void CirQueue<T>::print() {
		unsigned count = 0;
		for (auto ptr = head; ptr; ptr = ptr->next) {
			printf("\t	[%d]\t [%d]	\n", ptr->val, count++);
		}
	}

	template <typename T>
	std::tuple<T, bool>  CirQueue<T>::front()
	{
		if (!!head) {
			return { head->val, true };
		}
		return { 0, NULL };
	}

	template <typename T>
	unsigned CirQueue<T> ::size()
	{
		return curr_size;
	}

	template <typename T>
	bool CirQueue<T>::is_empty()
	{
		return head == NULL;
	}

	template <typename T>
	bool CirQueue<T>::contains(const T& value)
	{
		for (auto ptr = head; ptr; ptr = ptr->next) {
			if (ptr->val == value) {
				return true;
			}
		}
		return false;
	}

	template <typename T>
	void CirQueue<T>::enqueue(const T& value)
	{
		SLNode<T> *tmp, *node = new SLNode<T>{ value, NULL };

		if (curr_size >= cap) {
			tmp = head; 
			head = head->next;
			delete tmp;
		}
	
		if (!head) {
			head = node;
			tail = head;
		} else {
			tail->next = node;
			tail = node; 
		}
		curr_size += +(curr_size < cap);

	}

	template <typename T>
	std::tuple<T, bool>  CirQueue<T>::dequeue()
	{
		SLNode<T> *target; 
		T value;
		
		if (!!head) {
			target = head; 
			value = target->val;
			head = head->next;
			delete target; 
			return { value , true };
			--curr_size;
		}
		return { NULL, false };
	}

	template <typename T>
	bool CirQueue<T>::is_full()
	{
		return curr_size == cap;
	}

	template class CirQueue<unsigned>;
	template class CirQueue<signed>;

	unsigned weak_finger(unsigned weak_finger, unsigned usage_count)
	{
		unsigned curr_finger = 1, tick = 1, counter = 0, weak_finger_count = 0;
		
		if ((weak_finger < 1 && weak_finger > 5)) {
			return 0; 
		}
		
		while (!(weak_finger == curr_finger && weak_finger_count == usage_count)) {
			
			if (curr_finger == 1) {
				tick = 1;
			}

			if (curr_finger == 5) {
				tick = -1;
			}
			
			weak_finger_count += (curr_finger == weak_finger) ? 1: 0;
			curr_finger += tick;
			counter++;
		
		}
		return counter;
	}

}

File: chapter6/python/chapter6.py
from typing import NewType
import math 

class SLQueue():
  pass

class SLStack():
  pass 

SLQueue_t = NewType('SLQueue_t', SLQueue)
SLStack_t = NewType('SLStack_t', SLStack)

class SLNode:
  def __init__(self, value):
    self.val = value
    self.next = None 

class SLQueue:

  def __init__(self):
    self.head = None 
    self.tail = None

  def enqueue(self, val):
    node = SLNode(val) 
    if (self.head == None):
      self.head = node 
      self.tail = node 
    else:
      self.tail.next = node 
      self.tail = node
  
  def dequeue(self):
    node = self.head
    self.head = self.head.next
    if node == self.tail:
      self.tail = self.head
    return node.val
  
  @property
  def front(self):
    if self.head:
      return self.head.val
    else: 
      return None 
  
  def contains(self, value):
    runner = self.head 
    while runner:
      if runner.val == value:
        return True 
      runner = runner.next 
    return False 

  def is_empty(self):
    return True if not self.head else  False
  
  @property
  def size(self):
    count = 0  
    runner = self.head 
    while runner:
      count += 1
      runner = runner.next    
    return count 

  def compare_queues(self, a: SLQueue_t, b: SLQueue_t):
    size_a = a.size
    size_b = b.size
    
    if size_a != size_b:
      return False

    for i in range(0, size_a):
      if a.front != b.front:
        return False
      a.enqueue(a.dequeue())
      b.enqueue(b.dequeue())
    return True 
  
  def remove_min(self):
    min = None
    tmp = None 

    for i in range(0 , self.size):
      if min == None or self.front < min:
        min  = self.front
      self.enqueue(self.dequeue()) 
     
    for i in range(0 , self.size):
      tmp = self.dequeue()    
      if  tmp  != min:
        self.enqueue(tmp) 
  
  def interleave_queue(self):
    length = self.size
    mid  = math.floor (length  / 2) +  +(length % 2 == 1)
    tmp_queue = SLQueue()
    
    for i in range (0, length):
      if i < mid:
        self.enqueue(self.dequeue())
      else: 
        tmp_queue.enqueue(self.dequeue())
    
    length = self.size
    
    for i in range (0, length):
      self.enqueue(self.dequeue())
      if tmp_queue.front:
        self.enqueue(tmp_queue.dequeue())

  def display(self):
    runner = self.head
    if not runner:
      return None 

    while runner:
      print(runner.val)
      runner = runner.next
  
  def palindrome(self):
    stack = SLStack()
    result = True 

    while self.size:
      stack.push(self.front())
      self.enqueue(self.dequeue())
    while stack.size:
      result &= stack.top() == self.front()
      self.enqueue(self.dequeue())
    return result 

class ArrStack:
  def __init__(self):
    self.__collection = []
    self.size = 0
  
  def push(self, value):
    self.__collection = self.__collection + [value]
    self.size += 1

  def pop(self):
    result = None 
    if self.size:
      self.size -= 1
      result = self.__collection[self.size]
      self.__collection = self.__collection[0: self.size]
    return result
  
  def top(self):
    return self.__collection[0]
  
  def contains(self, value):
    return value in self.__collection
  
  def is_empty(self):
    return self.size == 0

  def display(self):
    print(self.__collection)

class SLStack:
  def __init__(self):
    self.head = None 
    self.tail = None 
    self.size = 0

  def top(self):
    return self.head.val

  def push(self, value):
    node = SLNode(value)
    if self.head == None:
      self.head = node
      self.tail = node 
    else:
      node.next = self.head 
      self.head = node 
    self.size += 1

  def pop(self):
    result = None 
    if self.head == None:
      return result  
    if self.head == self.tail:
      self.tail = None 
    result = self.head.val
    self.head = self.head.next 
    self.size -= 1
    return result 
  
  def contains(self, value):
    node = self.head
    while (node):
      if (node.val == value):
        return True
      node = node.next 
  
  def is_empty(self):
    return self.size == 0
  
  def remove_stack_min (self):
    queue = SLQueue()
    min_val = None  
    count = 0 
    data = None 

    while self.size:
      if min_val == None or self.top() < min_val:
        min_val = self.top()
        count += 1
      queue.enqueue(self.pop()) 
    
    while count:
      for i in range(0, count - 1):
        queue.enqueue(queue.dequeue())
      data = queue.dequeue()
      if (data != min_val):
        self.push(data)
      count -= 1  
       
  def display(self):
    runner = self.head 
    while runner:
      print(runner.val)
      runner = runner.next 


def compare_stacks(a: SLStack_t, b: SLStack_t):
  cmpr = True 
  length = None 

  stack = SLStack_t() 

  if a.size != b.size:
    return False 
  
  length = a.size

  for i in range (0, length):
      cmpr &= a.top == b.top
      stack.push(a.pop())
      stack.push(b.pop())
   
  for i in range (0, length):
     a.push(stack.pop())
     b.push(stack.pop())
  
  return cmpr

def copy_stack(stack: SLStack_t):
  queue = SLQueue()
  data = None 
  while stack.size:
    queue.enqueue(stack.pop())
  
  for i in range (queue.size - 1, -1, -1):
    for j in range (0, i):
      queue.enqueue(queue.dequeue())
    stack.push(queue.dequeue())

    

class Queue_Two_Stacks:
  def __init__(self):
    self.a = SLStack()
    self.b = SLStack()

  def push(self, value):
    while self.b.size:
     self.a.push(self.b.pop()) 
    self.a.push(value)
    while self.a.size:
      self.b.push(self.a.pop())
  
  def pop(self):
    result = None 
    
    while self.b.size:
      self.a.push(self.b.pop())
    
    result = self.a.pop()

    while self.a.size:
      self.b.push(self.a.pop())
  
    return result 


class StackQueueBase:
  def __init__(self):
    self._head = None 
    self._tail = None 
    self.size = 0

  def is_empty(self):
    return self.size == 0
  
  def contains(self, value):
    runner = self._head
    while runner:
      if runner.val == value:
        return True 
      runner = runner.next
    return False  

class SLQueue2(StackQueueBase):
  def __init__(self):
    super().__init__()

  def enqueue(self, value):
    node = SLNode(value)
    self.size += 1
    if self._head == None:
      self._head = node
      self.tail = node  
    else:
      self.tail.next = node
      self.tail = node 
  
  def dequeue(self):
    result = None
    if self.size:
      self.size -= 1
      result = self.head.val 
      if self.head == self.tail:
        self.tail = None 
      self.head = self.head.next  
    return result   

class SLStack2(StackQueueBase):
  def __init__(self):
    super().__init__()

  def push(self, value):
    node = SLNode(value)
    self.size += 1
    if self._head == None:
      self._head = node
      self._tail = node 
    else:
      node.next = self._head
      self._head = node 
  
  def pop (self):
    result = None 
    if self.size:
      if self._head == self._tail:
        self._tail = None 
      result = self._head.val
      self._head = self._head.next 
  

class Dequeue(SLStack2, SLQueue2):
  def __init__(self):
    super().__init__()
  
  def push_front(self, value):
    self.push(value)
  
  def pop_front(self):
    return self.pop()

  def push_back(self, value):
    self.enqueue(value)
  
  def pop_back(self):
    for i in range(0, self.size - 1):
      self.enqueue(self.dequeue())
    return self.pop()
  
  def front(self):
    return self._head.val

  def back(self):
    return self._tail.val
  
  def contains(self, value):
    return self.contains(value)

  def is_empty(self):
    return self.is_empty()

class CirQueue():
  def __init__(self, size = 10):
    self.head = None
    self.tail = None 
    self.capacity = size
    self.size = 0

  def front(self):
    return self.head.val
  
  def is_empty(self):
    return self.size == 0

  def is_full(self):
    return self.size == self.capacity 
  
  def contains(self, value):
    runner = self.head
    while runner:
      if value == runner.val:
        return True 
      runner = runner.next
    return False 
  
  def enqueue(self, value):
    if self.size < self.capcity:
      node = SLNode(value)
      if self.head == None:
        self.head = node 
        self.tail = node 
      else:
        self.tail.next = node 
        self.tail = node 
      self.size += 1

  def dequeue(self):
    result = None 
    if self.size:
      if (self.tail == self.head):
        self.tail = None 
      result = self.head.val
      self.head = self.head.next
      self.size -= 1
    return result 
  
  def grow (self, size):
    runner = None 
    count = 0 
    if (size < self.capacity and  size > 0):
      while runner and count < size :
        if count == 0:
          runner = self.head
        else:
          runner = runner.next 
        count += 1
      runner.next = None # break chain
    self.capacity = size 

def reorder_absolute_queue(queue: SLQueue_t):
  stack = SLStack()
  data = None 

  while queue.size:
    data = queue.dequeue()
    while stack.size > 0 and stack.top() < data:
      queue.enqueue(stack.pop())
    stack.push(data)
  while stack.size:
    queue.enqueue(stack.pop())

def partition(stack: SLStack_t):
  queue = SLQueue()
  count = 0

  while stack.size:
    queue.enqueue(stack.pop())
    count += 1
  
  for i in range (0, count):
    data = queue.dequeue()
    if data >= 0:
      stack.push(data)
    else:
      queue.enqueue(data)
    
    while queue.size:
      stack.push(queue.dequeue())

def switch_pairs(stack: SLStack_t):
  queue = SLQueue()
  count = 0

  while stack.size:
    queue.enqueue(stack.pop())
    count += 1

  while count > 0:
    rot = count -2
    for i in range (0, rot):
      queue.enqueue(queue.dequeue())
    
    if queue.size:
      stack.push(queue.dequeue())
    if queue.size:
      stack.push(queue.dequeue())
    count -= 2
  while queue.size:
    stack.push(queue.dequeue())
 
def is_sorted(stack:SLStack_t):
  stack2 = SLStack()
  sorted_stack = True
  count = 0 

  while stack.size:
    if stack2.size:
      sorted_stack &= stack.top() < stack2.top()
    stack2.push(stack.pop())
  return sorted_stack

def mirror(stack: SLStack_t):
  stack_tmp = SLStack()
  count = 0
  data = None 

  while stack.size:
    stack_tmp.push(stack.pop())

  while (stack_tmp.size):
    data = stack_tmp.pop()
    count = 0
    
    while stack.size:
      count += 1
      stack_tmp.push(stack.pop())
    
    stack.push(data)
    while count > 0:
      stack.push(stack_tmp.pop())
      count -= 1
    stack.push(data)






